libcache/cache.c

```diff
@@ -0,0 +1,460 @@
+	uint64_t setMask;
+	uint64_t offMask;
+
+	uint64_t offBitNum;
+	uint64_t setBitsNum;
+	uint64_t tagBitsNum;
+
+	cache_readCb_t readCb;
+	cache_writeCb_t writeCb;
+};
+
+
+static void cache_setInit(cacheset_t *set)
+{
+	set->timestamps = NULL;
+	memset(set->tags, 0, sizeof(set->tags));
+	memset(set->lines, 0, sizeof(set->lines));
+
+	set->count = 0;
+}
+
+
+/* Generates mask of type uint64_t with numBits set to 1 */
+static uint64_t cache_genMask(int numBits)
+{
+	return ((uint64_t)1 << numBits) - (uint64_t)1;
+}
+
+
+cachectx_t *cache_init(size_t size, size_t lineSize, cache_writeCb_t writeCb, cache_readCb_t readCb)
+{
+	int i = 0, j = 0;
+	cacheset_t *set = NULL;
+
+	cachectx_t *cache = malloc(sizeof(cachectx_t));
+
+	if (cache == NULL) {
+		return NULL;
+	}
+
+	if (size % lineSize != 0 || size % LIBCACHE_NUM_WAYS != 0) {
+		fprintf(stderr, "Invalid size parameters: size in bytes must be multiple of lineSize and multiple of %d\n", LIBCACHE_NUM_WAYS);
+		free(cache);
+		return NULL;
+	}
+
+	cache->lineSize = lineSize;
+	cache->numLines = size / cache->lineSize;
+
+	cache->numSets = cache->numLines / LIBCACHE_NUM_WAYS;
+
+	cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+	if (cache->sets == NULL) {
+		free(cache);
+		return NULL;
+	}
+
+	for (; i < cache->numSets; ++i) {
+		cache_setInit(&cache->sets[i]);
+	}
+
+	cache->offBitNum = log2(cache->lineSize);
+	cache->setBitsNum = log2(cache->numSets);
+	cache->tagBitsNum = LIBCACHE_ADDR_WIDTH - cache->setBitsNum - cache->offBitNum;
+
+	cache->tagMask = cache_genMask(cache->tagBitsNum);
+	cache->setMask = cache_genMask(cache->setBitsNum);
+	cache->offMask = cache_genMask(cache->offBitNum);
+
+	cache->readCb = readCb;
+	cache->writeCb = writeCb;
+
+	return cache;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j;
+
+	for (i = 0; i < cache->numSets; ++i) {
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			if (IS_VALID(cache->sets[i].lines[j].flags)) {
+				free(cache->sets[i].lines[j].data);
+				CLEAR_VALID(cache->sets[i].lines[j].flags);
+			}
+		}
+	}
+	free(cache->sets);
+
+	free(cache);
+
+	return 0;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const uint64_t offset, size_t count, int policy)
+{
+	int i, j;
+	ssize_t written = 0;
+	cacheline_t *temp = NULL, **linePtr = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
```

Comment:
I suggest
```suggestion
		for (i = 0; i < set->count; ++i) {
```

Reply:
This required moving the incrementation of ``set->count`` before that loop so that the loop could be entered at all. Anyway, I've applied your suggestion. This will omit a few operations at the beginning.